### PR TITLE
integrate knpMenu into sonata Admin

### PR DESCRIPTION
**Pull request by amine26tz**

### CODE REVIEW

File: Admin/Pool.php [hunk 1]
```diff
@@ -330,4 +333,51 @@ public function getOption($name, $default = null)
...
+    /**
+     * Create and return a knp MenuItem
+     *
+     * @return ItemInterface
+     */
+    public function getMenu()
+    {
+        $menuFactory = new MenuFactory();
+        $menu = $menuFactory->createItem('root')
+                            ->setExtra('request', $this->container->get('request'))
```
Comment (Bladrak) [inline Admin/Pool.php]: Indent

**Comment by Bladrak** [comment]
Missing doc (update configuration, changelog, bc breaks, ...)

**Comment by rande** [comment]
Can you add a bit of documentation ?

**Comment by amine26tz** [comment]
ping @pulzarraider @rande

File: Admin/Pool.php [hunk 2]
```diff
@@ -330,4 +333,52 @@ public function getOption($name, $default = null)
...
+    /**
+     * Create and return a knp MenuItem
+     *
+     * @return ItemInterface
+     */
+    public function getMenu()
+    {
+        $menuFactory = new MenuFactory();
+        $menu = $menuFactory
+                    ->createItem('root')
```
Comment (Bladrak) [inline Admin/Pool.php]: Indent

Comment (amine26tz) [inline Admin/Pool.php]: @Bladrak done

**Comment by Bladrak** [comment]
Seems good to me :)

File: Resources/doc/reference/knp_menu.rst [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    # Default configuration for "SonataAdminBundle"
+    sonata_admin:
+        dashboard:
+            groups:
+                news:
+                    label:                ~
+                    label_catalogue:      ~
+                    items:
+                        - sonata.news.admin.post
+                        - route: blog_home
```
Comment (rande) [inline Resources/doc/reference/knp_menu.rst]: Does the code handle route with parameter ?

File: Admin/Pool.php [hunk 4]
```diff
@@ -330,4 +333,52 @@ public function getOption($name, $default = null)
...
+
+            foreach ($group['items'] as $item) {
+                if (array_key_exists('admin', $item) && $item['admin'] != null) {
+                    $admin             = $this->getInstance($item['admin']);
+                    $label             = $admin->getLabel();
+                    $route             = $admin->generateUrl('list');
+                    $translationDomain = $admin->getTranslationDomain();
+                } else {
+                    $label             = $item['label'];
+                    $route             = $this->container->get('router')->generate($item['route']);;
```
Comment (rande) [inline Admin/Pool.php]: you need to handle route with parameters

**Comment by amine26tz** [comment]
ping @rande

**Comment by eko** [comment]
This is a cool feature, it will be useful to a lot of people :) :+1:

File: Resources/views/standard_layout.html.twig [hunk 5]
```diff
@@ -205,47 +205,7 @@ file that was distributed with this source code.
...
-                                                                {% if admin.hasroute('list') and admin.isGranted('LIST') %}
-                                                                    <li{% if app.request.get('_sonata_admin') == admin.code %} class="active"{% endif %}><a href="{{ admin.generateUrl('list')}}"><i class="fa fa-angle-double-right"></i> {{ admin.label|trans({}, admin.translationdomain) }}</a></li>
-                                                                {% endif %}
-                                                            {% endfor %}
-                                                        </ul>
-                                                    </li>
-                                                {% endif %}
-                                            {% endfor %}
-                                        </ul>
+                                        {{ knp_menu_render(admin_pool.menu, {'template' : 'SonataAdminBundle:Menu:sonata_menu.html.twig'}) }}
```
Comment (rande) [inline Resources/views/standard_layout.html.twig]: Cannot we make the SonataAdminBundle:Menu:sonata_menu.html.twig configurable like other template ?

**Comment by amine26tz** [comment]
done @rande

File: Admin/Pool.php [hunk 6]
```diff
@@ -330,4 +333,52 @@ public function getOption($name, $default = null)
 
         return $default;
     }
+
+    /**
+     * Create and return a knp MenuItem
+     *
+     * @return ItemInterface
+     */
+    public function getMenu()
```
Comment (rande) [inline Admin/Pool.php]: Can we move this code into a twig helper ?

File: Resources/views/Menu/sonata_menu.html.twig [hunk 7]
```diff
@@ -0,0 +1,136 @@
+{% macro attributes(attributes) %}
```
Comment (hason) [inline Resources/views/Menu/sonata_menu.html.twig]: This template is incompatible with "knplabs/knp-menu >= 2.0" see https://github.com/KnpLabs/KnpMenu/releases/tag/v2.0.0-alpha2 

You should extend ``knp_menu.html.twig`` template that is compatible with KnpMenu ``~1.0`` and ``~2.0`` (please do a test):

```twig
{% extends 'knp_menu.html.twig' %}

{% block root %}
    {%- set listAttributes = item.childrenAttributes|merge({'class': 'sidebar-menu'}) %}
    {%- set request = item.getExtra('request') %}
    {{- block('list') -}}
{% endblock %}

{% block item %}
{%- if item.displayed %}
    {#- check role of the group #}
    {%- set display = (item.getExtra('roles') is empty or is_granted('ROLE_SUPER_ADMIN') ) %}
        {%- for role in item.getExtra('roles') if not display %}
            {%- set display = is_granted(role) %}
        {%- endfor %}
    {%- endif %}
{% endif %}

{%- if item.displayed and display|default %}
    {%- set active = false %}
    {%- if item.getExtra('admin') is not empty and if item.getExtra('admin').hasroute('list') and item.getExtra('admin').isGranted('LIST') and request.get('_sonata_admin') == item.getExtra('admin').code %}
         {%- set active = true %}
    {%- elseif item.route is defined and request.get('_route') == item.route %}
        {%- set active = true %}
    {%- else %}
        {%- for child in item.children if not active %}
            {%- if child.getExtra('admin') is not empty and child.getExtra('admin').hasroute('list') and child.getExtra('admin').isGranted('LIST') and request.get('_sonata_admin') == child.getExtra('admin').code %}
                {%- set active = true %}
            {%-  elseif child.route is defined and request.get('_route') == child.route %}
                {%- set active = true %}
            {%- endif %}
        {%- endfor %}
    {%- endif %}

    {%- if item.level == 1%}
        {%- do item.setAttribute('class', (item.attribute('class')~' treeview')|trim) %}
    {%- endif %}
    {%- if active %}
        {%- do item.setAttribute('class', (item.attribute('class')~' active')|trim) %}
        {%- do item.setChildrenAttribute('class', (item.childrenAttribute('class')~' active')|trim) %}
    {%- endif %}

    {%- do item.setChildrenAttribute('class', (item.childrenAttribute('class')~' treeview-menu')|trim) %}
    {{ parent() }}
{% endif %}
{% endblock %}

{% block linkElement %}
{% spaceless %}
    {% set translation_domain = item.getExtra('translationdomain', 'messages') %}
    {% set icon = item.level > 1 ? '<i class="fa fa-angle-double-right"></i>' : '' %}
    {{ parent() }}
{% endspaceless %}
{% endblock %}

{% block spanElement %}
{% spaceless %}
    <a href="#">
        {% set translation_domain = item.attribute('label_catalogue') %} 
        {% set icon item.attribute('icon')|default ? item.attribute('icon') : '' %}
        {{ parent() }}
        <i class="fa pull-right fa-angle-left"></i>
    </a>
{% endspaceless %}
{% endblock %}

{% block label %}{{ icon|default|raw }}{% if options.allow_safe_labels and item.getExtra('safe_label', false) %}{{ item.label|raw }}{% else %}{{ item.label|trans({}, translation_domain|default('messages')) }}{% endif %}{% endblock %}
```

File: Resources/views/Menu/sonata_menu.html.twig [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    {% endfor %}
+    {# building the class of the item #}
+    {%- set classes = item.attribute('class') is not empty ? [item.attribute('class')] : [] %}
+    {%- if item.level == 1%}
+        {%- set classes = classes|merge(['treeview']) %}
+    {%- endif %}
+    {% if active %}
+        {%- set classes = classes|merge(['active']) %}
+    {% endif %}
+    {%- if item.current %}
```
Comment (hason) [inline Resources/views/Menu/sonata_menu.html.twig]: Incompatible with KnpMenu > 2.0 https://github.com/KnpLabs/KnpMenu/blob/master/src/Knp/Menu/Resources/views/knp_menu.html.twig#L55

File: Resources/views/Menu/sonata_menu.html.twig [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        {%- set classes = classes|merge([options.currentClass]) %}
+    {%- elseif item.currentAncestor %}
```
Comment (hason) [inline Resources/views/Menu/sonata_menu.html.twig]: Incompatible with KnpMenu >= 2.0 https://github.com/KnpLabs/KnpMenu/blob/master/src/Knp/Menu/Resources/views/knp_menu.html.twig#L57

File: Resources/views/Menu/sonata_menu.html.twig [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    {% if item.getExtra('admin') is not empty %}
+        {% if item.getExtra('admin').hasroute('list') and item.getExtra('admin').isGranted('LIST') and request.get('_sonata_admin') == item.getExtra('admin').code %}
+            {% set active = true %}
+        {% endif %}
+    {% else %}
+        {%  if item.route is defined and request.get('_route') == item.route %}
+            {% set active = true %}
+        {% endif %}
+    {% endif %}
+    {{ block('item', {'active': active}) }}
```
Comment (hason) [inline Resources/views/Menu/sonata_menu.html.twig]: This is currently unsupported statement, see https://github.com/twigphp/Twig/pull/1433

**Comment by amine26tz** [comment]
thanx @hason for your recomendation, i've update the menu template with some minor modification.

**Comment by amine26tz** [comment]
done @rande

File: Twig/Extension/SonataAdminExtension.php [hunk 11]
```diff
@@ -60,12 +70,23 @@ public function initRuntime(\Twig_Environment $environment)
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+
+    public function getFunctions()
+    {
+        return array(
+            'sonata_knp_menu' => new \Twig_Function_Method($this, 'getKnpMenu')
```
Comment (rande) [inline Twig/Extension/SonataAdminExtension.php]: rename => ``sonata_knp_build_menu``

File: Twig/Extension/SonataAdminExtension.php [hunk 12]
```diff
@@ -343,4 +364,54 @@ public function getXEditableType($type)
...
     }
+
+    /**
+     * Get KnpMenu
+     *
+     * @param Request $request
+     *
+     * @return ItemInterface
+     */
+    public function getKnpMenu(Request $request)
```
Comment (rande) [inline Twig/Extension/SonataAdminExtension.php]: you should support for empty request => ``$request = null``

**Comment by amine26tz** [comment]
ping @rande

**Pull request merged**